src/core/algorithms/ind/faida/inclusion_testing/hyperloglog.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+
+/**
+ * The content of this file is based on the HLL implementation made by
+ * Hideaki Ohno (https://github.com/hideo55). Modified by alexandrsmirn.
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+#include "immintrin.h"
+
+#define HLL_HASH_SEED 313
+
+#if defined(__has_builtin) && (defined(__GNUC__) || defined(__clang__))
+
+#define _GET_CLZ(x, b) (uint8_t)std::min(b, ::__builtin_clz(x)) + 1
+
+#else
+
+inline uint8_t _get_leading_zero_count(uint32_t x, uint8_t b) {
+
+#if defined (_MSC_VER)
+    uint32_t leading_zero_len = 32;
+    ::_BitScanReverse(&leading_zero_len, x);
+    --leading_zero_len;
+    return std::min(b, (uint8_t)leading_zero_len);
+#else
+    uint8_t v = 1;
+    while (v <= b && !(x & 0x80000000)) {
+        v++;
+        x <<= 1;
+    }
+    return v;
+#endif
+
+}
+#define _GET_CLZ(x, b) _get_leading_zero_count(x, b)
+#endif /* defined(__GNUC__) */
+
+namespace hll {
+
+static const double pow_2_32 = 4294967296.0;       ///< 2^32
+static const double neg_pow_2_32 = -4294967296.0;  ///< -(2^32)
+
+/** @class HyperLogLog
+ *  @brief Implement of 'HyperLogLog' estimate cardinality algorithm
+ */
+class HyperLogLog {
+public:
+    /**
+     * Constructor
+     *
+     * @param[in] b bit width (register size will be 2 to the b power).
+     *            This value must be in the range[4,30].Default value is 4.
+     *
+     * @exception std::invalid_argument the argument is out of range.
+     */
+    HyperLogLog(uint8_t b = 4) noexcept(false) : b_(b), m_(1 << b), M_(m_, 0) {
+        if (b < 4 || 30 < b) {
+            throw std::invalid_argument("bit width must be in the range [4,30]");
+        }
+
+        double alpha;
+        switch (m_) {
+            case 16:
+                alpha = 0.673;
+                break;
+            case 32:
+                alpha = 0.697;
+                break;
+            case 64:
+                alpha = 0.709;
+                break;
+            default:
+                alpha = 0.7213 / (1.0 + 1.079 / m_);
+                break;
+        }
+        alphaMM_ = alpha * m_ * m_;
+    }
+
+    void add_hash_32(uint32_t hash) {
+        uint32_t index = hash >> (32 - b_);
+        uint8_t rank = _GET_CLZ((hash << b_), 32 - b_);
+        if (rank > M_[index]) {
+            M_[index] = rank;
+        }
+    }
+
+    int zcf(size_t hash, int prec) {
+        size_t mask = ~(((1 << prec) - 1) << (64 - prec));
+        return _lzcnt_u64(hash & mask) - (uint8_t) prec;
+    }
+
+    void add_hash(size_t value_hash) {
+        int index = (int) (value_hash >> (8 * sizeof(size_t) - b_));
+        int rank = _lzcnt_u64((value_hash << b_) | ((1 << (b_ - 1)) + 1)) + 1;
+        // int rank = zcf(value_hash, b_) + 1;
```

Comment:
Here I do not have enough competence to assess the correctness of this approach.
I would prefer to use a macro with a comment explaining its meaning.

Reply:
I do not think that the marco is needed (and do not know how a macro can be used here). This is just a line I did not use but did not want to just delete. In fact, it is mostly for me, in case in the future a problem occurs.